### ISSUE TITLE
Semantic Segmentation - `y_true` is changed during training when passed to custom metric

### ISSUE THREAD

**Issue by nHeidelberg** [keras-team/keras#12273]
While working on a semantic segmentation task, I needed to implement a custom metric instead of using accuracy. I noticed that the values are weird, so I decided to just print the maximum of my y_true tensor.  My assumption is that `y_true ` only contains the labels values.

**Information**: 
- Keras version 2.2.4 with Tensorflow backend ( v 1.12.0)
- I have a mask (which is my target/y_true) of shape (batchSize, imgHeight, imgWidth, imgRegions) = (1,512,512,2) which only contains values 0 or 1 (I already check that my mask doesn't have wrong values)
- Softmax is used, currently I have two classes (background and object of interest).

**For debugging:**
- No Data Augmentation was Used
- I used batch size of 1
- No class weighting 
- I use as my metric:
`def TP(y_true, y_pred):`
` return K.max(y_true[:,:,:,1]) + K.max(y_true[:,:,:,0])`
- In the training_generator.py file in Keras, line 154. I print the max value of the mask when it is fetched from the generator:
`if len(generator_output) == 2:`
`x, y = generator_output`
`import numpy as np`
`print(' Deb ', np.max(y[:,:,:,0]) + np.max(y[:,:,:,1]))`
`sample_weight = None`
The value is always correct (i.e the value is a whole number).
- I also tried summing the `y_pred ` tensor. It gives me a sum that equals 512*512 + eps (very small error) as expected.
- I use a simple model:
![image](https://user-images.githubusercontent.com/23250250/52798148-5779fd80-3045-11e9-9a68-d00328db9c01.png)
- My training function:
![image](https://user-images.githubusercontent.com/23250250/52798233-8001f780-3045-11e9-9a09-7c77e45b92a6.png)

Here are the results of my training:
![image](https://user-images.githubusercontent.com/23250250/52798800-a5dbcc00-3046-11e9-9524-f38b7c530cb9.png)

Did I miss something on the way I use keras for this task?

**Edit**: I tried the same with a Sigmoid and It looks like the problem persists.

**Comment by yymyymyym** [comment]
@nHeidelberg I have the similiar problem with you. I used different shape of data with a data generator, the shape of the y_true changed when training. Did you find any way to aviod the problem?
